rust/connlib/tunnel/src/tests.rs

```diff
@@ -218,175 +808,436 @@ impl ReferenceState {
     }
 }
 
-impl TunnelTest {
-    fn send_icmp_packet_client_to_gateway(
-        &mut self,
-        src: impl Into<IpAddr>,
-        dst: impl Into<IpAddr>,
-    ) {
-        let _maybe_transmit = self.client.encapsulate(
-            ip_packet::make::icmp_request_packet(src.into(), dst.into()),
-            self.now,
-        );
+#[derive(Clone)]
+struct SimNode<ID, S> {
+    id: ID,
+    state: S,
 
-        // TODO: Handle transmit (send to relay / gateway)
-    }
-}
+    ip4_socket: Option<SocketAddrV4>,
+    ip6_socket: Option<SocketAddrV6>,
 
-/// Generates a [`Transition`] that sends an ICMP packet to a random IP.
-///
-/// By chance, it could be that we pick a resource IP here.
-/// That is okay as our reference state machine checks separately whether we are pinging a resource here.
-fn icmp_to_random_ip() -> impl Strategy<Value = Transition> {
-    (any::<IpAddr>(), any::<IpAddr>())
-        .prop_map(|(src, dst)| Transition::SendICMPPacketToRandomIp { src, dst })
-}
+    tunnel_ip4: Ipv4Addr,
+    tunnel_ip6: Ipv6Addr,
 
-fn icmp_to_ipv4_cidr_resource() -> impl Strategy<Value = Transition> {
-    (any::<Ipv4Addr>(), any::<sample::Index>())
-        .prop_map(|(src, r_idx)| Transition::SendICMPPacketToIp4Resource { src, r_idx })
+    span: Span,
 }
 
-fn icmp_to_ipv6_cidr_resource() -> impl Strategy<Value = Transition> {
-    (any::<Ipv6Addr>(), any::<sample::Index>())
-        .prop_map(|(src, r_idx)| Transition::SendICMPPacketToIp6Resource { src, r_idx })
+#[derive(Clone)]
+struct SimRelay<S> {
+    id: RelayId,
+    state: S,
+
+    ip_stack: firezone_relay::IpStack,
+    allocations: HashSet<(AddressFamily, AllocationPort)>,
+    buffer: Vec<u8>,
+
+    span: Span,
 }
 
-/// Implementation of our reference state machine.
+/// Stub implementation of the portal.
```

Comment:
It would be really cool if I could use all this to test the Windows Clients in CI without having access to Docker / Elixir / a staging portal. but the SimPortal couldn't hook up to a real Client tunnel yet, right?

I think we had it as a long-term goal but it would take a while?

Reply:
These tests are running in Windows CI already! But they don't do any platform-specific stuff so unless the Rust compiler mis-compiles some of the state machine stuff, I'd be surprised to find a windows-specific bug with these tests.

Or what _exactly_ were you thinking of running in CI here?

Reply:
I mean I'd like to boot up the Client, open the GUI stuff, open wintun and the virtual network interface, and exercise all the platform-specific code with only a fake Portal and fake Gateway, since those don't run easily in Windows and are hard to hook into the runners.

Like sort of tell connlib "I know there is no portal but just connect everything anyway"